rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,82 @@
+
+## Details
+
+Teleport will now store a single cached version of the client tools in the user's $HOME directory. tsh/tctl will execute this cached version when communicating with the Teleport Cloud cluster.
+
+Cloud will host an endpoint for Teleport client tools version discovery. tsh/tctl will check this endpoint to identify the current cloud-stable version of Teleport. If the cached version is behind the current cloud-stable version, the cache will be updated to match the current cloud-stable version. This will ensure Teleport client tools are always executed using the latest compatible version of Teleport when communicating with a Teleport Cloud cluster.
+
+For now this feature will be disabled for self-hosted users and will only be supported for Cloud users.
+
+### Version discovery
+
+Cloud will host the cloud-stable version of Teleport at https://updates.releases.teleport.dev/v1/stable/cloud/client-version. This endpoint will serve a Teleport version e.g. `v13.2.3`.
+
+The client tools are not using the same endpoint that agents use to discover the cloud-stable version of Teleport because having a separate channel allows us to rollback a botched client release without having to rollback the cluster or agents at the same time.
+
+The stable version should be updated after each Cloud Tenant upgrade. This should follow the same workflow as updating the cloud-stable agent version endpoint. The version should be updated soon after Cloud Tenant upgrades have been completed.
+
+Unlike the agent version discovery model, there will not be an endpoint to identify if a critical update is available. The reason for this is because updates are not executed in a scheduled update window. Teleport client tools will check this endpoint on every login and update accordingly if a new version is available.
+
+On every login, tsh will pull the current cloud-stable version of Teleport and store the value in the .tsh directory under `$HOME/.tsh/cloud-stable-version`.
+
+### Caching
+
+A single version of the Teleport client tools will be stored in the user's $HOME directory. Teleport already makes use of a .tsh directory for storing tsh config. Cached versions of tsh/tctl will also live within this directory under ` $HOME/.tsh/cache/<version>/{tsh,tctl}`. Whenever a new version of the client tools are available, the existing cached version will be replaced.
+
+On every login, tsh will compare the currently cached version of the client tools with the current stable version stored in `$HOME/.tsh/cloud-stable-version`. If the cached version is behind the current stable version, a new version of the client tools will be downloaded to replace the existing ones. When downloading the Teleport client tools, the binaries will be downloaded directly. We need to make sure to download a version that is compatible with the system.
```

Comment:
Just curious, what's the point of caching the version and checking against it, rather than just always checking against the version endpoint?

Reply:
Depending on the implementation, I wasn't sure how often we'd be checking the current cloud-stable version. I thought it might be beneficial to reduce the number of times we ping the endpoint by caching the version on login.

I don't have a strong opinion on needing to cache the value. We can just check against the endpoint every time.